adapters/repos/db/shard_write_batch_objects.go

```diff
@@ -0,0 +1,273 @@
+//  \ V  V /  __/ (_| |\ V /| | (_| | ||  __/
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2020 SeMI Technologies B.V. All rights reserved.
+//
+//  CONTACT: hello@semi.technology
+//
+
+package db
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/boltdb/bolt"
+	"github.com/go-openapi/strfmt"
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+	"github.com/semi-technologies/weaviate/adapters/repos/db/storobj"
+)
+
+// return value map[int]error gives the error for the index as it received it
+func (s *Shard) putObjectBatch(ctx context.Context,
+	objects []*storobj.Object) map[int]error {
+	return newObjectsBatcher(s).Objects(ctx, objects)
+}
+
+// objectsBatcher is a helper type wrapping around an underlying shard that can
+// execute objects batch operations on a shard (as opposed to references batch
+// operations)
+type objectsBatcher struct {
+	sync.Mutex
+	shard      *Shard
+	statuses   map[strfmt.UUID]objectInsertStatus
+	errs       map[int]error
+	duplicates map[int]struct{}
+	objects    []*storobj.Object
+}
+
+func newObjectsBatcher(s *Shard) *objectsBatcher {
+	return &objectsBatcher{shard: s}
+}
+
+// Objects imports the specified objects in parallel in a batch-fashion
+func (b *objectsBatcher) Objects(ctx context.Context,
+	objects []*storobj.Object) map[int]error {
+	beforeBatch := time.Now()
+	defer b.shard.metrics.BatchObject(beforeBatch, len(objects))
+
+	b.init(objects)
+	b.storeObjectsInObjectStore(ctx)
+	b.storeObjectsAdditionalStorage(ctx)
+	return b.errs
+}
+
+func (b *objectsBatcher) init(objects []*storobj.Object) {
+	b.objects = objects
+	b.statuses = map[strfmt.UUID]objectInsertStatus{}
+	b.errs = map[int]error{} // int represents original index
+	b.duplicates = findDuplicatesInBatchObjects(objects)
+}
+
+// storeObjectsInObjectStore performs all storage operations on the underlying
+// key/value store, this is they object-by-id store, the docID-lookup tables,
+// as well as all inverted indices.
+func (b *objectsBatcher) storeObjectsInObjectStore(ctx context.Context) {
+	maxPerTransaction := 30
+	beforeObjectStore := time.Now()
+	wg := &sync.WaitGroup{}
+	for i := 0; i < len(b.objects); i += maxPerTransaction {
+		end := i + maxPerTransaction
+		if end > len(b.objects) {
+			end = len(b.objects)
+		}
+
+		batch := b.objects[i:end]
+		wg.Add(1)
+		go func(i int, batch []*storobj.Object) {
+			defer wg.Done()
+			var affectedIndices []int
+			if err := b.shard.db.Batch(func(tx *bolt.Tx) error {
+				var err error
+				affectedIndices, err = b.storeSingleBatchInTx(ctx, tx, i, batch)
+				return err
+			}); err != nil {
+				b.setErrorsForIndices(err, affectedIndices)
+			}
+		}(i, batch)
+	}
+	wg.Wait()
+	b.shard.metrics.ObjectStore(beforeObjectStore)
+}
+
+func (b *objectsBatcher) setErrorsForIndices(err error, affectedIndices []int) {
+	b.Lock()
+	err = errors.Wrap(err, "bolt batch tx")
+	for _, affected := range affectedIndices {
+		b.errs[affected] = err
+	}
+	b.Unlock()
```

Comment:
Just a minor suggestion, would change that to `defer b.Ulock()` and add that code just after `b.Lock()`

Reply:
Great point. Previously this was part of a massive method, so `defer` would have deferred this for really long, but now that it's so small the `defer b.Unlock()` at the beginning would have the desired effect 👍